src/microbe_stage/MicrobeAI.cs

```diff
@@ -369,44 +396,94 @@ private void RunAndTumble(Random random)
         {
             MoveWithRandomTurn(2.5f, 3.0f, random);
         }
 
         // If there isn't any food here, it's a good idea to keep moving
-        if (compoundDifference == 0 && random.Next(0, 10) < 5)
+        if (Math.Abs(gradientValue) <= differenceDetectionThreshold && random.Next(0, 10) < 5)
         {
             MoveWithRandomTurn(0.0f, 0.4f, random);
         }
 
-        // If positive last step you gained compounds, so let's stick around
-        if (compoundDifference > 0)
+        // If positive last step you gained compounds, so let's move toward the source
+        if (gradientValue > differenceDetectionThreshold)
         {
-            // There's a decent chance to turn most of the way around
+            // There's a decent chance to turn by 90° to explore gradient
+            // 180° is useless since previous position let you absorb less compounds already
             if (random.Next(0, 10) < 4)
             {
-                MoveWithRandomTurn(0.0f, 3.0f, random);
+                MoveWithRandomTurn(0.0f, 1.5f, random);
             }
         }
     }
 
+    /// <summary>
+    ///   Prioritizing compounds that are stored in lesser quantities.
+    ///   If ATP-producing compounds are low (less than half storage capacities),
+    ///   non ATP-related compounds are discarded.
+    ///   Updates compoundsSearchWeights class dictionary.
+    /// </summary>
+    private void ComputeCompoundsSearchWeights()
+    {
+        IEnumerable<Compound> usefulCompounds = microbe.TotalAbsorbedCompounds.Keys;
+
+        // If this microbe lacks vital compounds don't bother with ammonia and phosphate
+        if (microbe.TotalAbsorbedCompounds.Keys.Any(
+            compound => IsVitalCompound(compound) &&
+                microbe.Compounds.GetCompoundAmount(compound) < 0.5f * microbe.Compounds.Capacity))
+        {
+            usefulCompounds = usefulCompounds.Where(x => x != ammonia && x != phosphates);
+        }
+
+        compoundsSearchWeights.Clear();
+        foreach (var compound in usefulCompounds)
+        {
+            // The priority of a compound is inversely proportional to its availability
+            // Should be tweaked with consumption
+            var compoundPriority = 1 - microbe.Compounds.GetCompoundAmount(compound) / microbe.Compounds.Capacity;
+
+            compoundsSearchWeights.Add(compound, compoundPriority);
+        }
+    }
+
+    /// <summary>
+    ///   Tells if a compound is vital to this microbe.
+    ///   Vital compounds are* direct* ATP producers
+    ///   TODO: what is used here is a shortcut linked to the current game state:
+    ///     such compounds could be used for other processes in future versions
+    ///   TODO: currently, vital compounds are not necessarily useful (e.g. no use of iron)!
+    /// </summary>
+    private bool IsVitalCompound(Compound compound)
+    {
+        return compound == glucose || compound == iron;
```

Comment:
why didn't you write ` || (compound == iron && microbe.Compounds.IsUseful(iron))` here?

Reply:
Hm, the thing is, at this point I don't need to know whether it is useful to the microbe because it was filtered out already if not. So I just kept vital independant from useful, although in practice, I agree there is little reason for a compound to be vital but not useful. 

I'm a bit torn about changing it. As of now, this feels like a static method that wasn't defined as such, so I think I'll add some usefulness check indeed.

Note that this isn't limited to iron (glucose can be useless as well), by the way.